tiledb/sm/rest/curl.cc

```diff
@@ -180,21 +180,64 @@ static int buffer_list_seek_callback(
   return CURL_SEEKFUNC_FAIL;
 }
 
+/**
+ * This callback function gets called by libcurl as soon as a header has been
+ * received. libcurl buffers headers and delivers only "full" headers, one by
+ * one, to this callback. This callback should return the number of bytes
+ * actually taken care of if that number differs from the number passed to
+ * your callback function, it signals an error condition to the library
+ *
+ * @param res_data points to the delivered data
+ * @param size the size of that data is size multiplied with nmemb
+ * @param count number of bytes actually taken care of
+ * @param userdata the userdata argument is set with CURLOPT_HEADERDATA
+ * @return
+ */
+size_t OnReceiveData(
+    void* res_data, size_t size, size_t count, void* userdata) {
+  //  size_t length = size * count;
+  size_t index = size * count;
+  auto* buffer = static_cast<char*>(res_data);
+
+  //  while (index < length) {
+  //    char* temp = buffer + index;
+  //    if ((temp[0] == '\r') || (temp[0] == '\n'))
+  //      break;
+  //    index++;
+  //  }
+
+  std::string str(buffer, index);
+  size_t pos = str.find(": ");
+  auto* pmHeader = static_cast<HeaderCbData*>(userdata);
+  if (pos != std::string::npos) {
+    if (str.substr(0, pos) == constants::redirection_header_key) {
```

Comment:
Experimentally, I found the header to be lower case: `x-tiledb-full-region-domain: us-east-1.aws.api.tiledb.com`. This will fail because `constants::redirection_header_key` is upper-case. We should do a case-insensitive comparison here.

I did the following:
1. Change the constant to lower-case. `std::string redirection_header_key = "x-tiledb-full-region-domain";`
2. Converted the header key to lowercase with `std::transform` and `::tolower`.

Putting it all together, here's my implementation:
```
size_t OnReceiveData(
    void* res_data, size_t size, size_t count, void* userdata) {
  const size_t header_length = size * count;
  char* const header_buffer = static_cast<char*>(res_data);
  HeaderCbData* const pmHeader = static_cast<HeaderCbData*>(userdata);

  std::string header(header_buffer, header_length);

  const size_t header_key_end_pos = header.find(": ");
  if (header_key_end_pos != std::string::npos) {
    // Fetch the header key and ensure the header is lowercase.
    std::string header_key = header.substr(0, header_key_end_pos);
    std::transform(
        header_key.begin(), header_key.end(), header_key.begin(), ::tolower);

    if (header_key == constants::redirection_header_key) {
      // Fetch the header value. Subtract 1 from the `header_length` to
      // remove the trailing new-line character.
      const std::string header_value =
          header.substr(header_key_end_pos + 2, header_length - 1);

      std::unique_lock<std::mutex> rd_lck(*(pmHeader->redirect_uri_map_lock));
      pmHeader->redirect_uri_map->insert(
          std::make_pair(pmHeader->uri, header_value));
    }
  }

  return size * count;
}
```

Reply:
In my test newline character was not the only one. What about subtracting `/r`. Maybe we should go with 2?

Reply:
Right, it should have to account for both `/r` and `/n`. I had fixed this locally but after writing this comment, here's how I got it working:
https://github.com/TileDB-Inc/TileDB/blob/jpm/cache-redirection-wip/tiledb/sm/rest/curl.cc#L194